api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -250,16 +296,15 @@ private Map<DbWorkspace, Double> getFreeTierWorkspaceCostsFromBQ() {
                     + "GROUP BY project.id ORDER BY cost desc;")
             .build();
 
-    final Map<DbWorkspace, Double> workspaceCosts = new HashMap<>();
+    final Map<Long, Double> liveCostByWorkspace = new HashMap<>();
     for (FieldValueList tableRow : bigQueryService.executeQuery(queryConfig).getValues()) {
       final String project = tableRow.get("id").getStringValue();
-      if (workspacesIndexedByProject.containsKey(project)) {
-        workspaceCosts.put(
-            workspacesIndexedByProject.get(project), tableRow.get("cost").getDoubleValue());
+      if (workspaceByProject.containsKey(project)) {
```

Comment:
@calbach I have a question here. Instead of this if condition, can't we just add a where clause in the BigQuery to get only the projects with IDs in workspaceByProject map?
What I'm seeing is that we get all project from BigQuery, and we iterate over them even if workspaceByProject is actually empty...

Reply:
While this may be empty for you locally (as your database is probably close to empty), in other environments this map will contain 5k - 100k elements, which will be infeasible to include in the query.

Reply:
but we can still be smart by splitting the query to multiple smaller queries, for example if we have 5k workspace, we can split them in batches of 1k each and run multiple Big queries...wouldn't that be more efficient?

Reply:
I played a bit with the query and I put >100k elements in it, and it still worked. I am currently working on some more changes and I'll submit a new PR once I'm done

Reply:
To do this in a single query - I still don't think it's a great idea, I expect you're pushing pretty close to the query size limit. I wouldn't expect BigQuery to necessarily scale well in this direction, though maybe it will work for our purposes - I wouldn't be surprised to see complications down the road.  Multiple queries might be a good solution, especially if you're looking to break out the work into individual tasks.

As to whether it's more efficient - that depends what you mean. On the BigQuery side, I'd expect it's probably slightly more expensive to do it this way (more processing, less data returned). In terms of memory use on the server, it will use less memory, as a function of the amount of data in the billing view that's unrelated to the environment. So for local, you get large efficiency gains, and for test you probably get a relatively small improvement. staging/stable/preprod/prod would probably see moderate improvements to memory use.